posthog/temporal/batch_exports/temporary_file.py

```diff
@@ -0,0 +1,523 @@
+        )
+
+        self.default = default
+
+    def write(self, content: bytes) -> int:
+        """Write a single row of JSONL."""
+        n = self.batch_export_file.write(orjson.dumps(content, default=str) + b"\n")
+        return n
+
+    def _write_record_batch(self, record_batch: pa.RecordBatch) -> None:
+        """Write records to a temporary file as JSONL."""
+        for record in record_batch.to_pylist():
+            self.write(record)
+
+
+class CSVBatchExportWriter(BatchExportWriter):
+    """A `BatchExportWriter` for CSV format."""
+
+    def __init__(
+        self,
+        max_bytes: int,
+        flush_callable: FlushCallable,
+        field_names: collections.abc.Sequence[str],
+        extras_action: typing.Literal["raise", "ignore"] = "ignore",
+        delimiter: str = ",",
+        quote_char: str = '"',
+        escape_char: str | None = "\\",
+        line_terminator: str = "\n",
+        quoting=csv.QUOTE_NONE,
+        compression: str | None = None,
+    ):
+        super().__init__(
+            max_bytes=max_bytes,
+            flush_callable=flush_callable,
+            file_kwargs={"compression": compression},
+        )
+        self.field_names = field_names
+        self.extras_action: typing.Literal["raise", "ignore"] = extras_action
+        self.delimiter = delimiter
+        self.quote_char = quote_char
+        self.escape_char = escape_char
+        self.line_terminator = line_terminator
+        self.quoting = quoting
+
+        self._csv_writer: csv.DictWriter | None = None
+
+    @property
+    def csv_writer(self) -> csv.DictWriter:
+        if self._csv_writer is None:
+            self._csv_writer = csv.DictWriter(
+                self.batch_export_file,
+                fieldnames=self.field_names,
+                extrasaction=self.extras_action,
+                delimiter=self.delimiter,
+                quotechar=self.quote_char,
+                escapechar=self.escape_char,
+                quoting=self.quoting,
+                lineterminator=self.line_terminator,
+            )
+
+        return self._csv_writer
+
+    def _write_record_batch(self, record_batch: pa.RecordBatch) -> None:
+        """Write records to a temporary file as CSV."""
+        self.csv_writer.writerows(record_batch.to_pylist())
+
+
+class ParquetBatchExportWriter(BatchExportWriter):
+    """A `BatchExportWriter` for Apache Parquet format.
+
+    We utilize and wrap a `pyarrow.parquet.ParquetWriter` to do the actual writing. We default to their
+    defaults for most parameters; however this class could be extended with more attributes to pass along
+    to `pyarrow.parquet.ParquetWriter`.
+
+    See the pyarrow docs for more details on what parameters can the writer be configured with:
+    https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetWriter.html
+
+    In contrast to other writers, instead of us handling compression we let `pyarrow.parquet.ParquetWriter`
```

Comment:
Not so happy about the compression story, would like to define a more standard way of handling it, but pyarrow/Parquet has it's own way of doing things and I didn't want to mess with that.